### ISSUE TITLE
Is there a way to alter an expression so that it assumes that particular tracers are zero? 

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#3666]
I have a machine learning model like
```python
Model = Tuple[float, float]
```
And a pair of losses
```python
Losses = Tuple[float, float]
```
Different parameters in the model are supposed to use different loss terms.  So my training looks something like:
```python
def infer_and_train(model: Model) -> Tuple[Configuration, Model]:
    (c, l), f_vjp = vjp(configuration_and_losses, model)
    model_bars = [f_vjp(jnp.zeros_like(c), x) for x in [(1.0, 0.0), (0.0, 1.0)]]
    new_model = tuple(model[i] - model_bars[i][i] * learning_rate
                      for i in range(2))
    return c, new_model
```
This is necessary to prevent multiple calls to produce the configuration, which are expensive.

Would it be possible to add a JAX primitive like:
```python
def assume_zero(x: PyTree, y: PyTree) -> PyTree:
    "Returns a copy of x whereby all tracers in y are assumed to be zero."
```
Then, I could
- avoid the multiple calls to f_vjp, and
- embed the training of this model in a larger model without having to worry about the multiple vjp calls through the larger model being expensive: 
```python
@jit  # Jitting ensures that the arguments are transformed into tracers that can be used by assume_zero.
def g(f_vjp, c_bar, losses_bar):
    # Produce model_bar
    model_bar = f_vjp(c_bar, losses_bar)
    # Block gradients.
    return (assume_zero(model_bar[0], losses_bar[1]),
            assume_zero(model_bar[1], losses_bar[0]))


def infer_and_train(model: Model) -> Tuple[Configuration, Model]:
    (c, l), f_vjp = vjp(configuration_and_losses, model)
    model_bar = g(f_vjp, jnp.zeros_like(c), (1.0, 1.0))
    new_model = [model[i] - model_bars[i] * learning_rate
                 for i in range(2)]
    return c, new_model
```

**Comment by gnecula** [comment]
I do not quite understand the ask. I have added @mattjj, perhaps he can understand better.

It seems to me that you want to compute the Jacobian, and you want to avoid multiple calls to vjp. Perhaps `jax.jaxrev` can help here? It is the same in spirit to what you wrote except it uses `jax.vmap` to compute all VJPs in one batched computation. 

If the above suggestion does not help, I am wondering about the following:
```
def assume_zero(x: PyTree, y: PyTree) -> PyTree:
    "Returns a copy of x whereby all tracers in y are assumed to be zero."
```

By `tracer` there do you mean JAX tracer? That is an internal implementation detail, and it depends on the context of each transformation. Can you specify that this transformation in a different way?

**Comment by NeilGirdhar** [comment]
When I first learned about `vmap`, I thought it was mainly for batching training examples.   Now that I really learn it… Wow, JAX is a marvel of excellent design!  I love that `in_axes` can be any pytree!

Yes, you are right (of course) that I can apply `vmap` to `f_vjp` to process the different loss cotangents `[(1.0, 0.0), (0.0, 1.0)]` in parallel.  And then I can select the values I need from the resulting model cotangents.

Also, looking closely, I think `jacrev` actually might be a perfect fit if I make the configuration into an auxiliary output?  It's not clear from the documentation whether auxiliary output is a pytree or static output, but if it's a pytree, then this would be perfect for my case.

Whether I choose the `vmap` or `jacrev` solution, I could wrap the whole thing in a function, and discard the values I don't need and then I assume that your XLA compiler will elide unnecessary computations, which was the point of this feature request.  So, I can close this.  Thanks for finding two elegant solutions to my problem!

**Comment by jekbradbury** [comment]
The "auxiliary output" allowed by several autodiff API functions can be an arbitrary pytree, not just static data.

**Comment by NeilGirdhar** [comment]
@jekbradbury 

After some testing, it appears that it is static data and not a pytree.  The following code leaks a tracer:
```python
from jax import vjp, jit


class C:
    def __init__(self, x):
        self.x = x

    def __repr__(self):
        return f"C({self.x})"


def f(x):
    return 2 * x, C(3 * x)


@jit
def h():
    y, f_vjp, z = vjp(f, 2.5, has_aux=True)
    x_bar, = f_vjp(1.0)

    print(x_bar, y, z)  # z contains a tracer

    return z.x


h()  # jax.core.UnexpectedTracerError!!
```

It should probably be documented that `has_aux` output is static?

**Comment by NeilGirdhar** [comment]
@jekbradbury Cool!
